linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala

```diff
@@ -2100,141 +2017,123 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
-    fb ++= ctx.stringPool.getConstantStringInstr("length")
-    fb += Call(genFunctionID.jsSelect)
-    fb += Call(genFunctionID.unbox(IntRef))
-    fb += LocalTee(lengthsLenLocal)
-
-    // lengthsValues := array.new<i32Array> lengthsLen
-    fb += ArrayNewDefault(genTypeID.i32Array)
-    fb += LocalSet(lengthsValuesLocal)
-
-    // i := 0
-    fb += I32Const(0)
-    fb += LocalSet(iLocal)
-
-    // while (i != lengthsLen)
-    fb.whileLoop() {
-      fb += LocalGet(iLocal)
-      fb += LocalGet(lengthsLenLocal)
-      fb += I32Ne
-    } {
-      // lengthsValue[i] := lengths[i] (where the rhs is a JS field access)
+    // Check negative array size
+    if (semantics.negativeArraySizes != CheckedBehavior.Unchecked) {
+      fb += LocalGet(lengthParam)
+      fb += I32Const(0)
+      fb += I32LtS
+      fb.ifThen() {
+        fb += LocalGet(lengthParam)
+        fb += Call(genFunctionID.throwNegativeArraySizeException)
+        fb += Unreachable
+      }
+    }
 
-      fb += LocalGet(lengthsValuesLocal)
-      fb += LocalGet(iLocal)
+    // componentTypeData := jlClass.data
+    fb += LocalGet(jlClassParam)
+    fb += StructGet(genTypeID.ClassStruct, genFieldID.classData)
+    fb += LocalTee(componentTypeDataLocal)
 
-      fb += LocalGet(lengthsParam)
-      fb += LocalGet(iLocal)
-      fb += RefI31
-      fb += Call(genFunctionID.jsSelect)
-      fb += Call(genFunctionID.unbox(IntRef))
+    // Load the vtable and itables of the ArrayClass instance we will create
+    fb += I32Const(1)
+    fb += Call(genFunctionID.arrayTypeData) // vtable
+    fb += GlobalGet(genGlobalID.arrayClassITable) // itables
 
-      fb += ArraySet(genTypeID.i32Array)
+    // Load the length
+    fb += LocalGet(lengthParam)
 
-      // i += 1
-      fb += LocalGet(iLocal)
-      fb += I32Const(1)
-      fb += I32Add
-      fb += LocalSet(iLocal)
+    // switch (componentTypeData.kind)
+    val switchClauseSig = FunctionType(
+      List(arrayTypeDataType, RefType(genTypeID.itables), Int32),
+      List(RefType(genTypeID.ObjectStruct))
+    )
+    fb.switch(switchClauseSig) { () =>
+      // scrutinee
+      fb += LocalGet(componentTypeDataLocal)
+      fb += StructGet(genTypeID.typeData, genFieldID.typeData.kind)
+    }(
+      // case KindPrim => array.new_default underlyingPrimArray; struct.new PrimArray
+      primRefsWithArrayTypes.map { case (primRef, kind) =>
+        List(kind) -> { () =>
+          val arrayTypeRef = ArrayTypeRef(primRef, 1)
+          fb += ArrayNewDefault(genTypeID.underlyingOf(arrayTypeRef))
+          fb += StructNew(genTypeID.forArrayClass(arrayTypeRef))
+          () // required for correct type inference
+        }
+      }: _*
+    ) { () =>
+      // case _ => array.new_default anyrefArray; struct.new ObjectArray
+      val arrayTypeRef = ArrayTypeRef(ClassRef(ObjectClass), 1)
+      fb += ArrayNewDefault(genTypeID.underlyingOf(arrayTypeRef))
+      fb += StructNew(genTypeID.forArrayClass(arrayTypeRef))
```

Comment:
Regarding complexity of `Class_newArray`: I feel this is actually more complex now, because `NewArray` and `Class_newArray` are not aligned in what they do.

If `lenghts` was an array here, we could simply get the type data and call `newArrayObject`. We probably wouldn't even need a core helper for it.

Reply:
Sure, but part of the appeal of the switch is to make the `newInstance(length)` faster by not having to allocate an array, because it's the common case (like ... the other one is never used in practice 🙄).